File: app/src/main/java/com/gameautopilot/app/core/RateLimiter.kt
package com.gameautopilot.app.core

class RateLimiter(@Volatile var maxPerMinute: Int) {
    private val timestamps = ArrayDeque<Long>()

    @Synchronized
    fun tryAcquire(now: Long = System.currentTimeMillis()): Boolean {
        val windowStart = now - 60_000L
        while (timestamps.isNotEmpty() && timestamps.first() < windowStart) {
            timestamps.removeFirst()
        }
        if (timestamps.size >= maxPerMinute.coerceAtLeast(1)) return false
        timestamps.addLast(now)
        return true
    }

    @Synchronized
    fun reset() {
        timestamps.clear()
    }
}

File: mac/Sources/GameAutopilotMac/Action/RateLimiter.swift
import Foundation

/// Sliding 60-second-window action cap. Mirrors core/RateLimiter.kt.
final class RateLimiter {
    var maxPerMinute: Int
    private var timestamps: [TimeInterval] = []
    private let lock = NSLock()

    init(maxPerMinute: Int) {
        self.maxPerMinute = maxPerMinute
    }

    @discardableResult
    func tryAcquire(now: TimeInterval = Date().timeIntervalSince1970) -> Bool {
        lock.lock(); defer { lock.unlock() }
        let windowStart = now - 60.0
        while let first = timestamps.first, first < windowStart {
            timestamps.removeFirst()
        }
        if timestamps.count >= max(maxPerMinute, 1) { return false }
        timestamps.append(now)
        return true
    }

    func reset() {
        lock.lock(); defer { lock.unlock() }
        timestamps.removeAll()
    }
}
